src/lambda_toolshed/papillon.cljc

```diff
@@ -68,46 +65,69 @@
 
 (defn- move
   "Move to the next interceptor in the interceptor chain within `ctx`.  Returns
-  a tuple of the resulting context, the current interceptor and a descriptive
-  tag of the move operation.  Returns nil if the chain has been consumed."
-  [{::keys [queue stack stage trace] :as ctx}]
+  a three-tuple of the resulting context, the current interceptor and the
+  resulting stage keyword.  Returns nil if the chain has been consumed."
+  [{::keys [queue stack stage error handled?] :as ctx}]
   (case stage
     :enter (if-let [ix (peek queue)]
-             (let [tag [(identify ix) stage]
-                   ctx (cond-> ctx
-                         true (update ::queue pop)
-                         true (update ::stack conj ix)
-                         trace (update ::trace conj tag))]
-               [ctx ix tag])
-             (-> ctx (assoc ::stage :leave) move))
-    (:error :leave) (if-let [ix (peek stack)]
-                      (let [tag [(identify ix) stage]
-                            ctx (cond-> ctx
-                                  true (update ::stack pop)
-                                  trace (update ::trace conj tag))]
-                        [ctx ix tag])
-                      nil)))
+             (let [ctx (-> ctx
+                           (update ::queue pop)
+                           (update ::stack conj ix))]
+               [ctx ix :enter])
+             (recur (assoc ctx ::stage (if error :error :leave))))
+    :leave (when-let [ix (peek stack)]
+             (let [ctx (-> ctx
+                           (assoc ::stage :final))]
+               [ctx ix :leave]))
+    :error (when-let [ix (peek stack)]
+             (let [ctx (-> ctx
+                           (assoc ::stage :final)
+                           (assoc ::handled? true))]
+               [ctx ix :error]))
+    :final (if (or (nil? error) handled?)
+             (if-let [ix (peek stack)]
+               (let [ctx (-> ctx
+                             (update ::stack pop)
+                             (dissoc ::handled?)
+                             (assoc ::stage (if error :error :leave)))]
+                 [ctx ix :final])
+               nil)
+             (recur (assoc ctx ::stage :error)))))
```

Comment:
@cch1 Am I reading this correctly is that the flow is: `:leave` -> `:final` -> `:error` -> `:final`? and `:final` does the check to see if there is an `:error` to handle, and then immediately jumps back out to `:error` if so?

Reply:
No.  The flow is `:leave` -maybe-> `:error` -> `:final`.  At this point in the code, the "current" interceptor is finally (!!) popped off the stack and the stage is the stage at which the subsequent interceptor will start.  IOW:

`B :leave` -> `B :final` -> `A :leave` 
OR
`B :error` -> `B :final` -> `A :error` 
OR
`B :leave` -> `B :error` -> `B :final` -> `A :error` 

While it is possible for the `:final` stage to muck with the `::error` value and circumvent this "idiomatic" model, I discourage it in the README [here](https://github.com/lambda-toolshed/papillon/blob/002f5a3d549f1d56ecfc7644efaf2700bbd332c5/README.md?plain=1#L146).  This is analogous to the `finally` clause in Clojure not being able to change the return value.

Reply:
Those ^^ traces assume the `:error` stage does not clear `::error`.  If it does, then the subsequent (outer) `A` interceptor starts in the `:leave` stage of course.  For example:

`B :error` -> `B :final` -> `A :leave` 
OR
`B :leave` -> `B :error` -> `B :final` -> `A :leave`

Reply:
Example of that half-baked thought... Maybe something like this to "declare" the state transitions for the stage?

```clojure
(def stage-transitions 
  {:enter (fn [ctx]
          (cond
            (::error ctx) :error
            (seq (::queue ctx)) :enter
            :else :leave))

   :leave (fn [ctx]
            (cond
              (::error ctx) :error
              :else :final))
   :error (fn [ctx]
            (cond
              :else :final))

   :final (fn [ctx]
            (cond
              (::error ctx) :error
              :else :leave)) })


(defn ->next-stage [{:keys [stage] :as ctx}]
  ((stage stage-transitions) ctx))
```